### PR TITLE
feat: introduce package to standardise configuration

### PR DESCRIPTION
**Pull request by shreddedbacon**
This is a utility that aims to standardise the way we consume configurations when interacting with the Lagoon API via various tools.

The goal is to make it so that configuration for one tool can be used by another tool seamlessly as long as the configuration is in the expected location.

This utility is designed ONLY for the Lagoon API/UI/Auth interactions. If building a tool that has its own configurations, then a similar method of interacting with that tools configuration should leverage XDG in a similar manner but store the information in different files to remain independent of the core configuration.

# Considerations
* This is only for core configuration for tools, additional configurations should write their own configuration management that leverages concepts from this package without touching the core config files it generates.
* Migrations from existing configurations to this new format should be handled within those tools
  * the lagoon-cli will have a migration hook that can prompt a user to migrate their configuration if an old one is detected.
  * lagoon-sync should probably have something similar for the core configuration.
* `lagoon-sync` has additional configurations for handling syncs of different components (rsync/dbs, etc), these should be handled separately to the core configuration, either within lagoon-syncs own configuration utility, or an extension of this utility.

### CODE REVIEW

File: utils/config/context.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		} else {
+			return fmt.Errorf("context %s already exists", context.Name)
+		}
+	} else {
+		return fmt.Errorf("user %s not found, add user first", user)
+	}
+}
+
+// UpdateContext will update a context in the config if it exists
+func (c *Config) UpdateContext(context ContextConfig, user string) error {
```
Comment (bomoko) [inline utils/config/context.go:75]: Do you think there's any value in checking the incoming values of the context here in the same way we do when creating a new context?

Maybe just having a generic "isContextValid" type function shared between both?

File: utils/config/features.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+func featurePrefix(prefix, feature string) string {
+	return fmt.Sprintf("%s-%s", prefix, feature)
+}
+
+// SetGlobalFeature updates a feature within the global config
+func (c *Config) SetGlobalFeature(prefix, feature string, state bool) error {
+	if prefix == "" || feature == "" {
+		return fmt.Errorf("prefix or feature must not be empty")
```
Comment (bomoko) [inline utils/config/features.go:14]: this should read "prefix and feature must not be empty" - I think

File: utils/config/user.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		})
+		return nil
+	} else {
+		return fmt.Errorf("user %s already exists", user.Name)
+	}
+}
+
+// UpdateUser updates a user within the config
+func (c *Config) UpdateUser(user UserConfig) error {
+	if i, ok := c.checkUserExists(user.Name); ok {
```
Comment (bomoko) [inline utils/config/user.go:38]: same as with context - might be useful to check they haven't smashed the user's name here? Just check it's not empty (although this isn't as important)

**Review by bomoko — APPROVED** [review]
Great stuff. Couple comments, but happy to approve.

Looking forward to integrating this into sync, etc.

File: utils/config/context.go (hunk 1 above)
Comment (shreddedbacon) [reply utils/config/context.go:75]: Yep, makes sense

**Comment by shreddedbacon** [comment]
I've added some optional items to the config too just because they're possibly exposed by the [/.well-known/appspecific/sh.lagoon.discovery.json](https://github.com/uselagoon/lagoon/pull/3632) endpoint, and if we introduce discovery utilities, this information could be consumed into configuration easily.

**Pull request merged**